ext/dom/parentnode.c

```diff
@@ -0,0 +1,435 @@
+		documentNode = (xmlDoc *) contextNode;
+	} else {
+		documentNode = contextNode->doc;
+	}
+
+	fragment = xmlNewDocFragment(documentNode);
+
+	if (!fragment) {
+		return NULL;
+	}
+
+	stricterror = dom_get_strict_error(document);
+
+	for (i = 0; i < nodesc; i++) {
+		if (Z_TYPE(nodes[i]) == IS_OBJECT) {
+			ce = Z_OBJCE(nodes[i]);
+
+			if (instanceof_function(ce, dom_node_class_entry)) {
+				newNodeObj = Z_DOMOBJ_P(&nodes[i]);
+				newNode = dom_object_get_node(newNodeObj);
+
+				if (newNode->doc != documentNode) {
+					php_dom_throw_error(WRONG_DOCUMENT_ERR, stricterror);
+					return NULL;
+				}
+
+				if (newNode->parent != NULL) {
+					xmlUnlinkNode(newNode);
+				}
+
+				newNodeObj->document = document;
+				xmlSetTreeDoc(newNode, documentNode);
+
+				if (!xmlAddChild(fragment, newNode)) {
+					xmlFree(fragment);
+
+					php_dom_throw_error(HIERARCHY_REQUEST_ERR, stricterror);
+					return NULL;
+				}
+
+				continue;
+			} else {
+				xmlFree(fragment);
+
+				zend_type_error("Invalid argument type must be either DOMNode or string");
+				return NULL;
+			}
+		} else if (Z_TYPE(nodes[i]) == IS_STRING) {
+			newNode = xmlNewDocText(documentNode, (xmlChar *) Z_STRVAL(nodes[i]));
+
+			xmlSetTreeDoc(newNode, documentNode);
+
+			if (!xmlAddChild(fragment, newNode)) {
+				xmlFree(fragment);
+
+				return NULL;
+			}
+		} else {
+			xmlFree(fragment);
+
+			zend_type_error("Invalid argument type must be either DOMNode or string");
+
+			return NULL;
+		}
+	}
+
+	return fragment;
+}
+
+void dom_parent_node_append(dom_object *context, zval *nodes, int nodesc)
+{
+	xmlNode *parentNode = dom_object_get_node(context);
+	xmlNodePtr newchild, node, prevsib;
+	xmlNode *fragment = dom_zvals_to_fragment(context->document, parentNode, nodes, nodesc);
+
+	if (fragment == NULL) {
+		return;
+	}
+
+	newchild = fragment->children;
+	prevsib = parentNode->last;
+
+	if (newchild) {
+		if (prevsib != NULL) {
+			prevsib->next = newchild;
+		} else {
+			parentNode->children = newchild;
+		}
+
+		newchild->prev = prevsib;
+		parentNode->last = fragment->last;
+
+		node = newchild;
+		while (node != NULL) {
+			node->parent = parentNode;
+
+			if (node == fragment->last) {
+				break;
+			}
+			node = node->next;
+		}
+
+		fragment->children = NULL;
+		fragment->last = NULL;
+
+		dom_reconcile_ns(parentNode->doc, newchild);
+	}
+
+	xmlFree(fragment);
+}
+
+void dom_parent_node_prepend(dom_object *context, zval *nodes, int nodesc)
+{
+	xmlNode *parentNode = dom_object_get_node(context);
+
+	if (parentNode->children == NULL) {
+		dom_parent_node_append(context, nodes, nodesc);
+		return;
+	}
+
+	xmlNodePtr newchild, node, prevsib, nextsib;
+	xmlNode *fragment = dom_zvals_to_fragment(context->document, parentNode, nodes, nodesc);
+
+	if (fragment == NULL) {
+		return;
+	}
+
+	newchild = fragment->children;
+	prevsib = NULL;
+	nextsib = parentNode->children;
+
+	if (newchild) {
+		parentNode->children = newchild;
+		newchild->prev = prevsib;
```

Comment:
Both the prevsib variable and this assignment look unnecessary to me. prevsib is always NULL, and newchild->prev should also be NULL already.

Reply:
Yeah this was left over so that each method looks as equal to the previous ones so I can more easily find common things to refactor out later. They are too different though, except the parent node assignment so i cleaned this up now.